rts/motoko-rts/src/text_iter.rs

```diff
@@ -70,13 +66,13 @@ pub unsafe fn text_iter<M: Memory>(mem: &mut M, text: SkewedPtr) -> SkewedPtr {
 
 /// Returns whether the iterator is finished
 #[no_mangle]
-pub unsafe extern "C" fn text_iter_done(iter: SkewedPtr) -> u32 {
+pub unsafe extern "C" fn text_iter_done(iter: Value) -> u32 {
     let array = iter.as_array();
-    let pos = array.get(ITER_POS_IDX).0 >> 2;
+    let pos = array.get(ITER_POS_IDX).get_scalar();
     let blob = array.get(ITER_BLOB_IDX).as_blob();
     let todo = array.get(ITER_TODO_IDX);
 
-    if pos >= blob.len().0 as usize && todo == SkewedPtr(0) {
+    if pos >= blob.len().0 && matches!(todo.get(), PtrOrScalar::Scalar(0)) {
```

Comment:
Is the cost of this test really trivial or would it worth interoducing a specialized is_Zero function?

Reply:
I will check that. We also have a similar pattern in `continuation_table::table_initialized`.

Reply:
LLVM or rustc cannot optimize one of these

```
// 1
!matches!(TABLE.get(), PtrOrScalar::Scalar(0))
// 2
TABLE.get() == PtrOrScalar::Scalar(0)
```

into

```
TABLE.0 == 0
```

Which makes sense, because without knowing that `Scalar` will always have 0 in the least significant bit (an invariant we maintain in the constructors) it cannot assume that for `value >> 1` to be `0`, `value` needs to be `0`.

Doing this optimization manually reduces binary sizes and should improve performance. I will commit this change now.